### DISCUSSION TITLE
How would someone define an empty array?

### DISCUSSION THREAD

**Discussion by WORMSS** [colinhacks/zod#2201] [Q&A] [answered]
eg: 
```ts
type ObjectOfEmptyArrays = Record<string, []>;
```
How would someone do this in zod?
```ts
z.object({}).catchall(z.array(z.????));
```

**Comment by kostysh** [comment]
In typescript `[]` literally means `any[]`

so, you can use the following definition of schema:

```typescript
z.object({}).catchall(z.array(z.any()));
```

or

```typescript
z.object({}).catchall(z.array(z.uknown()));
```

**Reply by WORMSS** [reply]
The resulting `z.infer<>` would create
```ts
type ObjectOfEmptyArrays = Record<string, any[]>;
type ObjectOfEmptyArrays = Record<string, unknown[]>;
```
Which is not correct..
It would also not error on 
```ts
const shouldFail = { "test": ["I should be blank"] };
const result = z.object({}).catchall(z.array(z.unknown))).parse(shouldFail);
```

Though, that did give me the idea to test your statement.
```ts
const test1 = []; // Does indeed create any[];
```
But.. 
```ts
type Test2 = []; // Does not create any[];
```
![image](https://user-images.githubusercontent.com/1384537/225619181-3396f7c7-d256-4868-96a8-099e467c6faf.png)

Though, for some reason, this didn't seem to work for me.
```ts
z.object({}).catchall(z.array(z.never()));
```
but, it seems this did.. And it's giving me all the expected results in my code so far..
```ts
z.object({}).catchall(z.tuple([]));
```
Thank you for giving me the push to look even deeper..
Though, I am unsure this is obvious to everyone out there... 

Also not a massive fan on how you do catchall objects..

**Reply by kostysh** [reply]
```typescript
z.object({}).catchall(z.array(z.never()));`
```
definitely works as you want during the schema validation only 
but if you want to use inferred from schema type - you cannot, because this type restricts you from adding elements to the array.
You can workaround using `refine`:

```typescript
z.object({}).catchall(z.array(z.unknown()).refine((arr) => arr.length === 0, {
  message: 'Only empty arrays are supported,
}));
```

**Reply by kostysh** [reply]
here is the example:https://codesandbox.io/s/heuristic-paper-6qjbh5?file=/src/index.ts

**Reply by WORMSS** [reply]
Not adding to it later is exactly what I want. This is why `z.array(z.any()).length(0)` wouldn't work, because it only checked at parse time. After that it would let the user add stuff, which I didnt want.

**Reply by lionelmalletpro** [reply]
@kostysh  note that a simple/single apostrophe mark is missing at the end of the string message: 'Only empty arrays are supported',

**Comment by JacobWeisenburger** [comment, marked as answer]
I'm pretty sure what you are looking for is an empty tuple.

```ts
const schema = z.object( {} ).catchall( z.tuple( [] ) )
type Data = z.infer<typeof schema>
// type Data = {} & {
//     [ k: string ]: []
// }

console.log(
    schema.safeParse( { 'test': [] } ).success
) // true
console.log(
    schema.safeParse( { 'test': [ 'I should be blank' ] } ).success
) // false
```

**Reply by WORMSS** [reply]
https://github.com/colinhacks/zod/discussions/2201#discussioncomment-5333354

**Reply by JacobWeisenburger** [reply]
so....are you saying that you would like credit for the answer? or are you ok with my answer because it concurs with yours?

**Reply by WORMSS** [reply]
I'm saying it was already suggested and being used.

**Reply by JacobWeisenburger** [reply]
ok, I would like to mark this question as answered. Would you like me to credit you for the answer? or should I select mine?

**Reply by WORMSS** [reply]
Oh, didn't realise this was a Stack overflow knockoff. I thought it was just a discussion page.. Yeah, what ever you want.

**Reply by JacobWeisenburger** [reply]
> Oh, didn't realise this was a Stack overflow knockoff.

I don't know if it was your intent, but this statement is a bit rude. I'm trying to make sure as many questions get answered as possible. I'm more than happy to credit you with the answer. But if you continue being rude, I may have to take disciplinary actions. We are all working toward the goal of answering questions. There is no need to be rude.

So would you like me to credit you with the answer? or is my answer satisfactory?

**Reply by WORMSS** [reply]
Sorry you taken the comment as rude, but that was not my intention.. I wasn't saying your answer is a knockoff. I was saying this Github feature.. I honestly thought it was just a straight forward discussion page, I didn't even know "selecting answers" was "a thing" until you started talking about taking credit.. 
I had already selected your answer before you replied with your last comment.

**Reply by JacobWeisenburger** [reply]
Sorry for my misunderstanding. I'm not looking to get credit for things like what happens on Stack Overflow a lot, I simply want to make sure questions are answered. So as long as we have reached a satisfactory answer, I would like to make sure it is selected so other people know this has been answered when they search for this problem in the future. Thanks for selecting my answer. Again, I'm sorry for my misunderstanding of your comment.
